test/cmds/manifest/manifest.test.js

```diff
@@ -0,0 +1,318 @@
+const utils = require('../../../dist/src/utils/fs');
+const { manifest } = require('../../../dist/src/cmds/manifest/manifest');
+
+jest.mock('fs', () => ({
+  promises: {
+    writeFile: jest.fn(),
+    readFile: jest.fn(),
+  },
+}));
+
+const mockArgv = { dist: 'dist' };
+
+const getDefaultWeb3Wallet = () => {
+  return {
+    bundle: {
+      local: 'dist/bundle.js',
+      url: 'http://localhost:8081/dist/bundle.js',
+    },
+    initialPermissions: { alert: {} },
+  };
+};
+
+const getPackageJson = async (
+  web3Wallet = getDefaultWeb3Wallet(),
+  name = 'foo',
+  version = 1,
+  description = 'bar',
+  main = 'index.js',
+  repository = 'git',
+) => {
+  return {
+    name,
+    version,
+    description,
+    main,
+    web3Wallet,
+    repository,
+  };
+};
+
+global.snaps = {
+  verboseErrors: false,
+  suppressWarnings: false,
+  isWatching: false,
+};
+
+describe('manifest', () => {
+  describe('manifest function validates a snap package.json file', () => {
+
+    afterEach(() => {
+      jest.clearAllMocks();
+      jest.restoreAllMocks();
+    });
+
+    it('throws an error if there is no dist property', async () => {
+      const noDistArgv = {};
+      await expect(manifest(noDistArgv)).rejects.toThrow('Invalid params: must provide \'dist\'');
+    });
+
+    it('throws error if fs.readfile fails', async () => {
+      jest.spyOn(JSON, 'parse').mockImplementation(() => {
+        const err = new Error('file already exists');
+        err.code = 'ENOENT';
+        throw err;
+      });
+      const fsMock = jest.spyOn(fs.promises, 'readFile').mockImplementation();
+      await expect(manifest(mockArgv)).rejects.toThrow('Manifest error: Could not find package.json. Please ensure that you are running the command in the project root directory.');
+      expect(fsMock).toHaveBeenCalledTimes(1);
+    });
+
+    it('throws error if parse fails', async () => {
+      jest.spyOn(JSON, 'parse').mockImplementation(() => {
+        throw new Error('error');
+      });
+      const fsMock = jest.spyOn(fs.promises, 'readFile')
+        .mockImplementationOnce(async () => await getPackageJson());
+      await expect(manifest(mockArgv)).rejects.toThrow('Could not parse package.json');
+      expect(fsMock).toHaveBeenCalledTimes(1);
+    });
+
+    it('throws error if parsed json is invalid', async () => {
+      jest.spyOn(JSON, 'parse').mockImplementation(() => false);
+      const fsMock = jest.spyOn(fs.promises, 'readFile')
+        .mockImplementationOnce(async () => await getPackageJson());
+      await expect(manifest(mockArgv)).rejects.toThrow('Invalid parsed package.json: false');
+      expect(fsMock).toHaveBeenCalledTimes(1);
+    });
+
+    it('checks presence of required and recommended keys', async () => {
+      const badJSON = {};
+      jest.spyOn(JSON, 'parse').mockImplementation((value) => value);
+      jest.spyOn(console, 'warn').mockImplementation();
+      jest.spyOn(console, 'log').mockImplementation();
```

Comment:
These are repeated in almost every test case, and seem like a great candidate for a `beforeEach` hook.

Reply:
Only console error is repeated across 90%, so I will move the console error mock to a `beforeEach` hook.